src/repair.rs

```diff
@@ -54,26 +57,30 @@ pub enum RepairRequest {
 /// If well-formed, it thus contains the corresponding variant of [`RepairRequest`].
 #[derive(Clone, Debug, Serialize, Deserialize)]
 pub enum RepairResponse {
-    /// Response with the total number of slices for a specific block.
-    SliceCount(RepairRequest, usize),
+    /// Response with the last slice's Merkle root hash, plus corresponding proof.
+    LastSliceRoot(RepairRequest, SliceIndex, Hash, Vec<Hash>),
     /// Response with the Merkle root hash of a specific slice, plus corresponding proof.
     SliceRoot(RepairRequest, Hash, Vec<Hash>),
     /// Response with a specific shred.
     Shred(RepairRequest, Shred),
-    // TODO: remove or replace with variant that includes proof
-    Parent(RepairRequest, Slot, Hash),
 }
 
 /// Instance of double-Merkle based block repair protocol.
 pub struct Repair<N: Network> {
     blockstore: Arc<RwLock<Box<dyn Blockstore + Send + Sync>>>,
     pool: Arc<RwLock<Box<dyn Pool + Send + Sync>>>,
+
+    slice_roots: BTreeMap<(BlockId, SliceIndex), Hash>,
+    outstanding_requests: BTreeMap<Hash, RepairRequest>,
+    request_timeouts: BinaryHeap<(Instant, Hash)>,
+
     network: N,
```

Comment:
also something to consider doing in future PRs if we want to address it.

Currently, we are using the same port to handle requests and responses.  It might make sense to use separate ports for them so that we could prioritise replies over requests.

If we use separate ports, we can also have a separate struct for handling repair requests from the network.  That big of logic essentially has to look stuff up from blockstore and send replies.  Putting it in the same struct as where we repairing is not really necessary.

Reply:
true, these two parts could definitely be separated